### PR TITLE
feat(preact-project-gen): A Preact project generator

### PR DESCRIPTION
**Pull request by JayaKrishnaNamburu**

### CODE REVIEW

File: packages/teleport-project-packer-test/src/constants.ts [hunk 1]
```diff
@@ -7,3 +7,7 @@ export const REACT_NEXT_GITHUB_PROJECT = 'teleport-project-template-react-next'
 export const VUE_GITHUB_PROJECT = 'teleport-project-template-vue-basic'
 
 export const VUE_NUXT_GITHUB_PROJECT = 'teleport-project-template-vue-nuxt'
+
+export const PREACT_TEMPLATE_OWNER = 'JayaKrishnaNamburu'
```
Comment (JayaKrishnaNamburu) [inline packages/teleport-project-packer-test/src/constants.ts]: TODO: change the owner details, once the template repo is transferred.

File: packages/teleport-plugin-react-app-routing/src/index.ts [hunk 2]
```diff
@@ -31,8 +37,10 @@ export const createPlugin: ComponentPluginFactory<AppRoutingComponentConfig> = (
 
   const reactAppRoutingComponentPlugin: ComponentPlugin = async (structure) => {
     const { uidl, dependencies, options } = structure
+    // @ts-ignore-next-line
```
Comment (alexnm) [inline packages/teleport-plugin-react-app-routing/src/index.ts]: I would drop this in favor of explicit initialization for all the 3 values you are destructuring below:

```const { useFolderStructure, disableDOMInjection, flavour } = options.meta || { useFolderStructure: false, disableDOMInjection: false, ... }```

File: packages/teleport-plugin-react-app-routing/src/index.ts [hunk 3]
```diff
@@ -31,8 +37,10 @@ export const createPlugin: ComponentPluginFactory<AppRoutingComponentConfig> = (
...
+    const { useFolderStructure, disableDOMInjection, flavour } = options.meta || {}
 
-    registerRouterDeps(dependencies)
+    RouteDependencies[flavour](dependencies)
```
Comment (alexnm) [inline packages/teleport-plugin-react-app-routing/src/index.ts]: you can have an if statement here, later a switch if needed, it is a bit more clear, you don't have to shift attention to another place to check which possible flavors exist

File: packages/teleport-project-generator-preact-basic/README.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# `@teleporthq/teleport-project-generator-preact`
+
+> TODO: description
```
Comment (alexnm) [inline packages/teleport-project-generator-preact-basic/README.md]: fill this up with the same info you have in other packages

File: packages/teleport-plugin-react-app-routing/src/index.ts [hunk 5]
```diff
@@ -31,8 +37,10 @@ export const createPlugin: ComponentPluginFactory<AppRoutingComponentConfig> = (
 
   const reactAppRoutingComponentPlugin: ComponentPlugin = async (structure) => {
     const { uidl, dependencies, options } = structure
+    // @ts-ignore-next-line
+    const { useFolderStructure, disableDOMInjection, flavour } = options.meta || {}
```
Comment (alexnm) [inline packages/teleport-plugin-react-app-routing/src/index.ts]: I think you're not using disableDOMInjection anymore right?

File: packages/teleport-project-generator-react-basic/src/index.ts [hunk 6]
```diff
@@ -36,6 +36,9 @@ export const createReactBasicGenerator = () => {
       generator: routingComponentGenerator,
       path: ['src'],
       fileName: 'index',
+      metaDataOptions: {
+        flavour: 'react',
```
Comment (alexnm) [inline packages/teleport-project-generator-react-basic/src/index.ts]: can this be the default and not be specified?

File: packages/teleport-project-generator/src/types.ts [hunk 7]
```diff
@@ -4,19 +4,27 @@ export interface ProjectStrategy {
   components: {
     generator: ComponentGenerator
     path: string[]
+    metaDataOptions?: {
+      useFolderStructure?: boolean
```
Comment (alexnm) [inline packages/teleport-project-generator/src/types.ts]: I would rename this to: "createFolderForEachComponent"

File: packages/teleport-project-generator/src/index.ts [hunk 8]
```diff
@@ -74,8 +74,14 @@ export const createProjectGenerator = (strategy: ProjectStrategy): ProjectGenera
     // Handling pages, based on the conditionals in the root node
     for (const routeNode of routeNodes) {
       const { files, dependencies } = await createPage(routeNode, strategy, options)
+      const { metaDataOptions } = strategy.pages
 
-      injectFilesToPath(rootFolder, strategy.pages.path, files)
+      const path =
+        metaDataOptions && metaDataOptions.useFolderStructure
```
Comment (alexnm) [inline packages/teleport-project-generator/src/index.ts]: I would move this logic in a function whose name will explain what is happening here

File: packages/teleport-project-generator/src/index.ts [hunk 9]
```diff
@@ -74,8 +74,14 @@ export const createProjectGenerator = (strategy: ProjectStrategy): ProjectGenera
...
+          ? [...strategy.pages.path, ...[files[0].name]]
```
Comment (alexnm) [inline packages/teleport-project-generator/src/index.ts]: also, you can do [ ...strategy.pages.path, files[0].name ]

File: packages/teleport-plugin-react-app-routing/src/index.ts (hunk 5 above)
Comment (JayaKrishnaNamburu) [reply packages/teleport-plugin-react-app-routing/src/index.ts]: Yes, I removed it. I missed it here, will remove 👍

File: packages/teleport-project-generator-react-basic/src/index.ts (hunk 6 above)
Comment (JayaKrishnaNamburu) [reply packages/teleport-project-generator-react-basic/src/index.ts]: Yeah, but I felt, in furture, if we plan to use the same routing plugin with one more flavor then we don't need to do all checks and then placing react as a fallback

File: packages/teleport-project-generator-preact-basic/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+  const htmlFileGenerator = createPreactComponentGenerator()
+  htmlFileGenerator.addPostProcessor(prettierHTML)
+
+  const generator = createProjectGenerator({
+    components: {
+      generator: preactComponentGenerator,
+      path: ['src', 'components'],
+      metaDataOptions: {
+        useFolderStructure: true,
```
[the remark this answers is not in the document]
Comment (JayaKrishnaNamburu) [reply packages/teleport-project-generator-preact-basic/src/index.ts]: I am not sure, App generated with **cli** is following this convention

File: packages/teleport-plugin-react-app-routing/src/index.ts [hunk 11]
```diff
@@ -49,7 +52,9 @@ export const createPlugin: ComponentPluginFactory<AppRoutingComponentConfig> = (
 
       dependencies[componentName] = {
         type: 'local',
-        path: `${pageDependencyPrefix}${fileName}`,
+        path: useFolderStructure
+          ? `${pageDependencyPrefix}${dashCaseToCamelCase(fileName)}/${fileName}`
```
[the remark this answers is not in the document]
Comment (JayaKrishnaNamburu) [reply packages/teleport-plugin-react-app-routing/src/index.ts]: yes yes, i am still working on those 👍

**Pull request merged**